File: Colage/Core/Design/Components.swift
import SwiftUI

// MARK: - Primary Button
struct ColagePrimaryButton: View {
    @Environment(.themeColor) private var themeColor
    let title: String
    let action: () -> Void
    var isLoading: Bool = false
    var isDisabled: Bool = false

    var body: some View {
        Button(action: {
            UIImpactFeedbackGenerator(style: .medium).impactOccurred()
            action()
        }) {
            HStack(spacing: 8) {
                if isLoading {
                    ProgressView()
                        .tint(.white)
                }
                Text(title)
                    .font(ColageFonts.bodyBold)
            }
            .frame(maxWidth: .infinity)
            .frame(height: 56)
            .background(isDisabled ? ColageColors.surfaceElevated : themeColor)
            .foregroundStyle(isDisabled ? ColageColors.textTertiary : .white)
            .clipShape(RoundedRectangle(cornerRadius: 16))
        }
        .disabled(isDisabled || isLoading)
    }
}

// MARK: - Text Field
struct ColageTextField: View {
    @Environment(.themeColor) private var themeColor
    let placeholder: String
    @Binding var text: String
    var keyboardType: UIKeyboardType = .default
    var autocapitalization: TextInputAutocapitalization = .sentences

    var body: some View {
        TextField(placeholder, text: $text)
            .font(ColageFonts.body)
            .foregroundStyle(ColageColors.textPrimary)
            .padding(.horizontal, 20)
            .frame(height: 56)
            .background(ColageColors.surface)
            .clipShape(RoundedRectangle(cornerRadius: 16))
            .overlay(
                RoundedRectangle(cornerRadius: 16)
                    .strokeBorder(ColageColors.border, lineWidth: 1)
            )
            .keyboardType(keyboardType)
            .textInputAutocapitalization(autocapitalization)
    }
}

// MARK: - OTP Code Field
struct OTPCodeField: View {
    @Environment(.themeColor) private var themeColor
    @Binding var code: String
    let length: Int
    var onComplete: ((String) -> Void)? = nil

    var body: some View {
        HStack(spacing: 12) {
            ForEach(0..<length, id: \.self) { index in
                let char = index < code.count
                    ? String(code[code.index(code.startIndex, offsetBy: index)])
                    : ""
                Text(char)
                    .font(ColageFonts.title)
                    .frame(width: 48, height: 60)
                    .background(ColageColors.surface)
                    .clipShape(RoundedRectangle(cornerRadius: 12))
                    .overlay(
                        RoundedRectangle(cornerRadius: 12)
                            .strokeBorder(
                                index < code.count ? themeColor : ColageColors.border,
                                lineWidth: index < code.count ? 2 : 1
                            )
                    )
            }
        }
        // Tappable text field below the boxes
        TextField("Enter code", text: $code)
            .keyboardType(.numberPad)
            .textContentType(.oneTimeCode)
            .font(ColageFonts.mono)
            .foregroundStyle(ColageColors.textPrimary)
            .multilineTextAlignment(.center)
            .frame(height: 44)
            .background(ColageColors.surface)
            .clipShape(RoundedRectangle(cornerRadius: 10))
            .padding(.horizontal, 40)
            .padding(.top, 8)
            .onChange(of: code) { _, newValue in
                // Limit to length
                if newValue.count > length {
                    code = String(newValue.prefix(length))
                }
                // Filter non-digits
                code = code.filter { $0.isNumber }
                if code.count == length {
                    onComplete?(code)
                }
            }
    }
}

// MARK: - Circular Avatar
struct AvatarView: View {
    @Environment(.themeColor) private var themeColor
    let imageURL: String?
    let size: CGFloat
    var borderColor: Color = themeColor
    var showBorder: Bool = true

    var body: some View {
        Group {
            if let url = imageURL, let imageUrl = URL(string: url) {
                AsyncImage(url: imageUrl) { phase in
                    switch phase {
                    case .success(let image):
                        image.resizable().scaledToFill()
                    default:
                        placeholderAvatar
                    }
                }
            } else {
                placeholderAvatar
            }
        }
        .frame(width: size, height: size)
        .clipShape(Circle())
        .overlay(
            Circle()
                .strokeBorder(showBorder ? borderColor : .clear, lineWidth: 2)
        )
    }

    var initials: String? = nil

    private var placeholderAvatar: some View {
        ZStack {
            ColageColors.surfaceElevated
            if let initials, !initials.isEmpty {
                Text(initials)
                    .font(.system(size: size * 0.35, weight: .semibold, design: .rounded))
                    .foregroundStyle(themeColor)
            } else {
                Image(systemName: "person.fill")
                    .font(.system(size: size * 0.4))
                    .foregroundStyle(ColageColors.textTertiary)
            }
        }
    }
}

// MARK: - Visibility Toggle Button
struct VisibilityToggle: View {
    @Environment(.themeColor) private var themeColor
    @Binding var isVisible: Bool

    var body: some View {
        Button {
            isVisible.toggle()
        } label: {
            Image(systemName: isVisible ? "eye.fill" : "eye.slash.fill")
                .font(.system(size: 18, weight: .medium))
                .foregroundStyle(isVisible ? ColageColors.textPrimary : ColageColors.textTertiary)
                .frame(width: 40, height: 40)
                .background(ColageColors.surface.opacity(0.8))
                .clipShape(Circle())
        }
    }
}

// MARK: - Floor Picker
struct FloorPicker: View {
    @Environment(.themeColor) private var themeColor
    @Binding var selectedFloor: Int
    let floors: [Int]

    var body: some View {
        VStack(spacing: 4) {
            ForEach(floors.reversed(), id: \.self) { floor in
                Button {
                    selectedFloor = floor
                } label: {
                    Text(floorLabel(floor))
                        .font(ColageFonts.captionBold)
                        .foregroundStyle(
                            selectedFloor == floor
                                ? themeColor
                                : ColageColors.textSecondary
                        )
                        .frame(width: 36, height: 32)
                        .background(
                            selectedFloor == floor
                                ? themeColor.opacity(0.15)
                                : ColageColors.surface.opacity(0.8)
                        )
                        .clipShape(RoundedRectangle(cornerRadius: 8))
                }
            }
        }
        .padding(4)
        .background(ColageColors.surface.opacity(0.6))
        .clipShape(RoundedRectangle(cornerRadius: 12))
    }

    private func floorLabel(_ floor: Int) -> String {
        switch floor {
        case ..<0: return "B\(abs(floor))"
        default: return "\(floor)"
        }
    }
}

// MARK: - Segmented Mode Picker
struct DiscoveryModePicker: View {
    @Environment(.themeColor) private var themeColor
    @Binding var activeMode: AppState.DiscoveryMode

    var body: some View {
        HStack(spacing: 0) {
            ForEach(AppState.DiscoveryMode.allCases, id: \.self) { mode in
                Button {
                    withAnimation(.spring(response: 0.3, dampingFraction: 0.7)) {
                        activeMode = mode
                    }
                    UIImpactFeedbackGenerator(style: .light).impactOccurred()
                } label: {
                    Text(mode.rawValue)
                        .font(ColageFonts.captionBold)
                        .foregroundStyle(
                            activeMode == mode
                                ? ColageColors.textPrimary
                                : ColageColors.textSecondary
                        )
                        .frame(width: 64, height: 32)
                        .background(
                            activeMode == mode
                                ? themeColor
                                : Color.clear
                        )
                        .clipShape(RoundedRectangle(cornerRadius: 10))
                }
            }
        }
        .padding(4)
        .background(.ultraThinMaterial)
        .clipShape(RoundedRectangle(cornerRadius: 14))
    }
}

// MARK: - Distance Formatting
extension Double {
    /// Format distance in feet to a human-readable string
    var formattedDistance: String {
        if self < 50 { return "nearby" }
        if self < 5280 { return "\(Int(self)) ft" }
        return String(format: "%.1f mi", self / 5280.0)
    }
}

// MARK: - String Initials Extension
extension String {
    /// Returns up to 2 initials from a name, e.g. "Emma Wilson" → "EW"
    var initials: String {
        let parts = self.split(separator: " ").map { String($0) }
        if parts.count >= 2 {
            return String(parts[0].prefix(1) + parts[1].prefix(1)).uppercased()
        }
        return String(self.prefix(2)).uppercased()
    }
}

// MARK: - Progress Steps Indicator
struct OnboardingProgress: View {
    @Environment(.themeColor) private var themeColor
    let currentStep: Int
    let totalSteps: Int

    var body: some View {
        HStack(spacing: 4) {
            ForEach(0..<totalSteps, id: \.self) { step in
                RoundedRectangle(cornerRadius: 2)
                    .fill(step <= currentStep ? themeColor : ColageColors.border)
                    .frame(height: 3)
            }
        }
        .padding(.horizontal, 24)
    }
}

File: android/app/src/main/java/com/colageclub/colage/core/design/Components.kt
package com.colageclub.colage.core.design

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Person
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.focus.FocusRequester
import androidx.compose.ui.focus.focusRequester
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalFocusManager
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardCapitalization
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import coil.compose.AsyncImage
import coil.request.ImageRequest

// MARK: - Primary Button
@Composable
fun ColagePrimaryButton(
    title: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    isLoading: Boolean = false,
    isDisabled: Boolean = false
) {
    Button(
        onClick = { if (!isLoading && !isDisabled) onClick() },
        modifier = modifier
            .fillMaxWidth()
            .height(56.dp),
        enabled = !isDisabled && !isLoading,
        shape = RoundedCornerShape(16.dp),
        colors = ButtonDefaults.buttonColors(
            containerColor = LocalThemeColor.current,
            disabledContainerColor = ColageColors.SurfaceElevated
        ),
        contentPadding = PaddingValues(horizontal = 24.dp)
    ) {
        if (isLoading) {
            CircularProgressIndicator(
                modifier = Modifier.size(20.dp),
                color = Color.White,
                strokeWidth = 2.dp
            )
            Spacer(Modifier.width(8.dp))
        }
        Text(
            text = title,
            style = ColageFonts.BodyBold,
            color = if (isDisabled) ColageColors.TextTertiary else Color.White
        )
    }
}

// MARK: - Text Field
@Composable
fun ColageTextField(
    placeholder: String,
    value: String,
    onValueChange: (String) -> Unit,
    modifier: Modifier = Modifier,
    keyboardType: KeyboardType = KeyboardType.Text,
    capitalization: KeyboardCapitalization = KeyboardCapitalization.Sentences,
    imeAction: ImeAction = ImeAction.Done,
    onImeAction: () -> Unit = {}
) {
    val focusManager = LocalFocusManager.current
    BasicTextField(
        value = value,
        onValueChange = onValueChange,
        modifier = modifier
            .fillMaxWidth()
            .height(56.dp)
            .background(ColageColors.Surface, RoundedCornerShape(16.dp))
            .border(1.dp, ColageColors.Border, RoundedCornerShape(16.dp)),
        textStyle = ColageFonts.Body.copy(color = ColageColors.TextPrimary),
        cursorBrush = SolidColor(LocalThemeColor.current),
        keyboardOptions = KeyboardOptions(
            keyboardType = keyboardType,
            capitalization = capitalization,
            imeAction = imeAction
        ),
        keyboardActions = KeyboardActions(
            onDone = {
                focusManager.clearFocus()
                onImeAction()
            },
            onNext = { onImeAction() },
            onGo = { onImeAction() }
        ),
        singleLine = true,
        decorationBox = { innerTextField ->
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(horizontal = 20.dp),
                contentAlignment = Alignment.CenterStart
            ) {
                if (value.isEmpty()) {
                    Text(
                        text = placeholder,
                        style = ColageFonts.Body.copy(color = ColageColors.TextTertiary)
                    )
                }
                innerTextField()
            }
        }
    )
}

// MARK: - OTP Code Field
@Composable
fun OTPCodeField(
    code: String,
    onCodeChange: (String) -> Unit,
    length: Int = 6,
    onComplete: (String) -> Unit = {}
) {
    val focusRequester = remember { FocusRequester() }

    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        modifier = Modifier.semantics {
            contentDescription = "Verification code, ${code.length} of $length digits entered"
        }
    ) {
        // Visual boxes
        Row(
            horizontalArrangement = Arrangement.spacedBy(12.dp),
            modifier = Modifier
                .padding(horizontal = 24.dp)
                .clickable { focusRequester.requestFocus() }
        ) {
            repeat(length) { index ->
                val char = if (index < code.length) code[index].toString() else ""
                val isFilled = index < code.length
                Box(
                    modifier = Modifier
                        .size(width = 48.dp, height = 60.dp)
                        .background(ColageColors.Surface, RoundedCornerShape(12.dp))
                        .border(
                            width = if (isFilled) 2.dp else 1.dp,
                            color = if (isFilled) LocalThemeColor.current else ColageColors.Border,
                            shape = RoundedCornerShape(12.dp)
                        )
                        .clickable { focusRequester.requestFocus() },
                    contentAlignment = Alignment.Center
                ) {
                    Text(
                        text = char,
                        style = ColageFonts.Title.copy(color = ColageColors.TextPrimary),
                        textAlign = TextAlign.Center
                    )
                }
            }
        }

        // Hidden text field that captures input
        BasicTextField(
            value = code,
            onValueChange = { newValue ->
                val filtered = newValue.filter { it.isDigit() }.take(length)
                onCodeChange(filtered)
                if (filtered.length == length) {
                    onComplete(filtered)
                }
            },
            modifier = Modifier
                .size(0.dp)
                .alpha(0f)
                .focusRequester(focusRequester),
            keyboardOptions = KeyboardOptions(
                keyboardType = KeyboardType.NumberPassword,
                imeAction = ImeAction.Done
            ),
            textStyle = ColageFonts.Mono.copy(
                color = ColageColors.TextPrimary,
                textAlign = TextAlign.Center
            ),
            cursorBrush = SolidColor(LocalThemeColor.current),
            singleLine = true,
            decorationBox = { innerTextField ->
                Box(
                    modifier = Modifier
                        .fillMaxSize()
                        .padding(horizontal = 16.dp),
                    contentAlignment = Alignment.Center
                ) {
                    innerTextField()
                }
            }
        )
    }

    // Auto-focus removed — BringIntoViewRequester crashes on some devices
    // User taps the digit row to focus instead
}

// MARK: - Avatar View
@Composable
fun AvatarView(
    imageUrl: String?,
    size: Dp,
    modifier: Modifier = Modifier,
    borderColor: Color = LocalThemeColor.current,
    showBorder: Boolean = true,
    initials: String? = null,
    name: String? = null
) {
    Box(
        modifier = modifier
            .size(size)
            .clip(CircleShape)
            .then(
                if (showBorder) Modifier.border(2.dp, borderColor, CircleShape)
                else Modifier
            ),
        contentAlignment = Alignment.Center
    ) {
        if (imageUrl != null) {
            AsyncImage(
                model = ImageRequest.Builder(LocalContext.current)
                    .data(imageUrl)
                    .crossfade(true)
                    .build(),
                contentDescription = if (name != null) "Profile photo for $name" else "Profile photo",
                contentScale = ContentScale.Crop,
                modifier = Modifier.fillMaxSize()
            )
        } else {
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .background(ColageColors.SurfaceElevated),
                contentAlignment = Alignment.Center
            ) {
                if (!initials.isNullOrEmpty()) {
                    Text(
                        text = initials,
                        style = ColageFonts.Body.copy(
                            fontWeight = FontWeight.SemiBold,
                            fontSize = (size.value * 0.35).sp,
                            color = LocalThemeColor.current
                        )
                    )
                } else {
                    Icon(
                        imageVector = Icons.Default.Person,
                        contentDescription = null,
                        tint = ColageColors.TextTertiary,
                        modifier = Modifier.size(size * 0.4f)
                    )
                }
            }
        }
    }
}

// MARK: - Onboarding Progress
@Composable
fun OnboardingProgress(
    currentStep: Int,
    totalSteps: Int,
    modifier: Modifier = Modifier
) {
    Row(
        modifier = modifier
            .fillMaxWidth()
            .padding(horizontal = 24.dp),
        horizontalArrangement = Arrangement.spacedBy(4.dp)
    ) {
        repeat(totalSteps) { step ->
            Box(
                modifier = Modifier
                    .weight(1f)
                    .height(3.dp)
                    .background(
                        color = if (step <= currentStep) LocalThemeColor.current else ColageColors.Border,
                        shape = RoundedCornerShape(2.dp)
                    )
            )
        }
    }
}

// MARK: - String Extensions
fun String.initials(): String {
    val parts = this.trim().split(" ").filter { it.isNotEmpty() }
    return if (parts.size >= 2) {
        "${parts[0].first()}${parts[1].first()}".uppercase()
    } else {
        this.take(2).uppercase()
    }
}

// MARK: - Distance Formatting
fun Double.formattedDistance(): String {
    return when {
        this < 50 -> "nearby"
        this < 5280 -> "${this.toInt()} ft"
        else -> String.format("%.1f mi", this / 5280.0)
    }
}
